useReducer in calculator

jchavarri opened the pull request:
Test using `useReducer` in calculator

jchavarri committed:
useReducer in calculator

jchavarri commented:
Commenting in https://github.com/revery-ui/revery/pull/174#issuecomment-451256065 🙂

OhadRau commented:
This looks really good to me. I'll review this as soon as I'm done with lunch

OhadRau commented on examples/Calculator.re:
```diff
@@ -5,245 +5,341 @@ open Revery.UI.Components;
+      }
+  | ClearKeyPressed(ac) =>
+    ac ?
+      {...state, operator: `Nop, result: 0., display: ""} :
```
Shouldn't this also clear the `number` field? Otherwise, this'll only clear the current equation but leave the current value "typed in"

jchavarri replied:
Oh right, great catch! I thought the existing statement was an if/else but the reset of `curNum` was always happening. Fixed in https://github.com/OhadRau/revery/pull/1/commits/240b993fb18015795788506d4c0d4ec83a298475. 👍

OhadRau commented on examples/Calculator.re:
```diff
@@ -5,245 +5,341 @@ open Revery.UI.Components;
-        };
-        setCurNum("");
-      };
+let eval = (state, newOp) => {
```
Now that we have the reducer, maybe we can refactor `eval` into an action? Thoughts on this?

jchavarri replied:
I think `eval` wouldn't really fit an action-like role. My mental model for actions is "actions are like newspapers", in the sense that actions need to tell you something that happened in your app (generally user originated, or network originated). This talk has a really great overview on what are good candidates for actions: https://youtu.be/Dm9NgjR5Jn4?t=485

`eval` would fit more a "command" role. Because multiple actions (in this case 2) need to perform the evaluation, I think it makes sense to keep `eval` as just a helper function that encapsulates the common logic in the two actions reducers. There are plenty of ways to solve this though 😄 Do you think we should rethink the way `eval` and the current actions fit together?

jchavarri replied:
One rule of thumb for actions is to try to use the past tense to model them. So things like `KeyPressed` or `NetworkMessageReceived` would be good candidates for actions, but `ProcessQueue` or `CalculateDeviation` would not be as good (they are more "command like" and probably need to live in the reducer as helper functions). I hope that makes sense!

OhadRau reviewed (approved):
This looks great. I made two little comments but both are pretty minor things that I can look into before this gets merged. Thanks for taking the time to work on this!

jchavarri committed:
Reset `number` when clear key is pressed

jchavarri commented:
Thanks @OhadRau for the quick review! I fixed the issue with `ClearKeyPressed` handling, and added some thoughts about the `eval` treatment. Let me know what you think and feel free to merge and take over if you think this is g2g! 👍

OhadRau commented:
Great! Merging this now
